apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/context/util/CrossThreadRefContainer.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.agent.core.context.util;
+
+public class CrossThreadRefContainer {
```

Comment:
Could you explain why need this? Rather than use referred object directly?

Reply:
Previously, we use the `ResponseHolder` to capture the reference of `AbstractSpan` for async usage. Since an instance of ResponseHolder is set to the dynamic field of the enhanced object, as well as the ThreadLocal context, this can be used as a bridge for sharing span among `InvokeInterceptor.beforeMethod`, `InvokeInterceptor.afterMethod` and `AbstractMethodInterceptor`. The reason for the necessity of this bridge is that `InvokeInterceptor.beforeMethod` and `InvokeInterceptor.afterMethod` (specifically the `Consumer<SignalType> onFinally` parameter)  in the webflux context are not running in the same thread.

Since we've removed `ResponseHolder`, we need another container here. And this class must be in the agent jar, thus it is loaded by the `AppClassLoader`.

Reply:
Are you saying you need a reference to propagate something across before to after?

Reply:
If so, read this doc, we provide v2 plugin APIs for method context.

Reply:
Yes. It is a problem. I don't know who wrote this code before. I was just following the previous idea.

But I suppose probably we can still save the reference in the ThreadLocal. But in the `InvokeInterceptor.afterMethod`, we take it out first and make it effectively final and then refer to it in the lambda.

Reply:
We have v2 APIs added specifically for this case. @dmsolr is a member of SkyWalking PMC. He is the lead of these new APIs, and we have changed one plugin which has a better performance. You could sync with the master branch, then you could use v2 to make codes easier and better.

Reply:
> We have v2 APIs added specifically for this case. @dmsolr is a member of SkyWalking PMC. He is the lead of these new APIs, and we have changed one plugin which has a better performance. You could sync with the master branch, then you could use v2 to make codes easier and better.

Thanks for the information. I will try to adapt the Spring plugins with the new API. But I don't know whether I should/have to transform all spring plugins in this PR?

Besides, an additional e2e case have to be added for this particular issue. I will work on this.

Reply:
> But I don't know whether I should/have to transform all spring plugins in this PR?

You don't have to change all. V2 APIs are not for replacing V1, they are enhancing only. So, only change when the new feature helps. Make sense?

Reply:
> > But I don't know whether I should/have to transform all spring plugins in this PR?
> 
> You don't have to change all. V2 APIs are not for replacing V1, they are enhancing only. So, only change when the new feature helps. Make sense?

Great. I will have a look.